<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Interactive Visualization of RNA-seq Data Using a Principal Components Approach • pcaExplorer</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.3.7/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Interactive Visualization of RNA-seq Data Using a Principal Components Approach">
<meta property="og:description" content="This package provides functionality for interactive visualization 
    of RNA-seq datasets based on Principal Components Analysis. The methods provided
    allow for quick information extraction and effective data exploration. A Shiny
    application encapsulates the whole analysis.">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">pcaExplorer</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">2.9.4</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="articles/pcaExplorer.html">User guide</a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/upandrunning.html">Up and running with `r Biocpkg("pcaExplorer")`
</a>
    </li>
  </ul>
</li>
<li>
  <a href="news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/federicomarini/pcaExplorer">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="contents col-md-9">
<div id="pcaexplorer---interactive-exploration-of-principal-components-of-samples-and-genes-in-rna-seq-data" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#pcaexplorer---interactive-exploration-of-principal-components-of-samples-and-genes-in-rna-seq-data" class="anchor"></a>pcaExplorer - Interactive exploration of Principal Components of Samples and Genes in RNA-seq data</h1></div>
<div id="software-status" class="section level2">
<h2 class="hasAnchor">
<a href="#software-status" class="anchor"></a>Software status</h2>
<table class="table">
<thead><tr class="header">
<th align="center">Platforms</th>
<th align="center">OS</th>
<th align="center">R CMD check</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="center">Travis CI</td>
<td align="center">Linux</td>
<td align="center"><a href="https://travis-ci.org/federicomarini/pcaExplorer"><img src="https://travis-ci.org/federicomarini/pcaExplorer.svg?branch=master" alt="Travis CI build status"></a></td>
</tr>
<tr class="even">
<td align="center">Bioc (<a href="http://bioconductor.org/packages/devel/bioc/html/pcaExplorer.html"><em>devel</em></a>)</td>
<td align="center">Multiple</td>
<td align="center"><a href="http://bioconductor.org/checkResults/devel/bioc-LATEST/pcaExplorer"><img src="http://bioconductor.org/shields/build/devel/bioc/pcaExplorer.svg" alt="Bioconductor-devel Build Status"></a></td>
</tr>
<tr class="odd">
<td align="center">Bioc (<a href="http://bioconductor.org/packages/release/bioc/html/pcaExplorer.html"><em>release</em></a>)</td>
<td align="center">Multiple</td>
<td align="center"><a href="http://bioconductor.org/checkResults/release/bioc-LATEST/pcaExplorer"><img src="http://bioconductor.org/shields/build/release/bioc/pcaExplorer.svg" alt="Bioconductor-release Build Status"></a></td>
</tr>
</tbody>
</table>
<!-- 
[![Windows Build status](https://ci.appveyor.com/api/projects/status/github/federicomarini/pcaExplorer?svg=true)](https://ci.appveyor.com/project/federicomarini/pcaexplorer)
-->
<p><code>pcaExplorer</code> is a Bioconductor package containing a Shiny application for analyzing expression data in different conditions and experimental factors.</p>
<p>It is a general-purpose interactive companion tool for RNA-seq analysis, which guides the user in exploring the Principal Components of the data under inspection.</p>
<p><code>pcaExplorer</code> provides tools and functionality to detect outlier samples, genes that show particular patterns, and additionally provides a functional interpretation of the principal components for further quality assessment and hypothesis generation on the input data.</p>
<p>Moreover, a novel visualization approach is presented to simultaneously assess the effect of more than one experimental factor on the expression levels.</p>
<p>Thanks to its interactive/reactive design, it is designed to become a practical companion to any RNA-seq dataset analysis, making exploratory data analysis accessible also to the bench biologist, while providing additional insight also for the experienced data analyst.</p>
</div>
<div id="installation" class="section level2">
<h2 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h2>
<p><code>pcaExplorer</code> can be easily installed using <code><a href="https://www.rdocumentation.org/packages/BiocManager/topics/install">BiocManager::install()</a></code>:</p>
<pre><code>if (!requireNamespace("BiocManager", quietly=TRUE))
    install.packages("BiocManager")
BiocManager::install("pcaExplorer")</code></pre>
<p>or, optionally,</p>
<pre><code><a href="https://www.rdocumentation.org/packages/BiocManager/topics/install">BiocManager::install("federicomarini/pcaExplorer")
# or alternatively...
devtools::install_github("federicomarini/pcaExplorer")</a></code></pre>
</div>
<div id="quick-start" class="section level2">
<h2 class="hasAnchor">
<a href="#quick-start" class="anchor"></a>Quick start</h2>
<p>This command loads the <code>pcaExplorer</code> package</p>
<pre><code><a href="https://www.rdocumentation.org/packages/base/topics/library">library("pcaExplorer")</a></code></pre>
<p>The <code>pcaExplorer</code> app can be launched in different modes:</p>
<ul>
<li><p><code><a href="reference/pcaExplorer.html">pcaExplorer(dds = dds, rld = rld)</a></code>, where <code>dds</code> is a <code>DESeqDataSet</code> object and <code>rld</code> is a <code>DESeqTransform</code> object, which were created during an existing session for the analysis of an RNA-seq dataset with the <code>DESeq2</code> package</p></li>
<li><p><code><a href="reference/pcaExplorer.html">pcaExplorer(dds = dds)</a></code>, where <code>dds</code> is a <code>DESeqDataSet</code> object. The <code>rld</code> object is automatically computed upon launch.</p></li>
<li><p><code><a href="reference/pcaExplorer.html">pcaExplorer(countmatrix = countmatrix, coldata = coldata)</a></code>, where <code>countmatrix</code> is a count matrix, generated after assigning reads to features such as genes via tools such as <code>HTSeq-count</code> or <code>featureCounts</code>, and <code>coldata</code> is a data frame containing the experimental covariates of the experiments, such as condition, tissue, cell line, run batch and so on.</p></li>
<li><p><code><a href="reference/pcaExplorer.html">pcaExplorer()</a></code>, and then subsequently uploading the count matrix and the covariates data frame through the user interface. These files need to be formatted as tab separated files, which is a common format for storing such count values.</p></li>
</ul>
<p>Additional parameters and objects that can be provided to the main <code>pcaExplorer</code> function are:</p>
<ul>
<li><p><code>pca2go</code>, which is an object created by the <code>pca2go</code> function, which scans the genes with high loadings in each principal component and each direction, and looks for functions (such as GO Biological Processes) that are enriched above the background. The offline <code>pca2go</code> function is based on the routines and algorithms of the <code>topGO</code> package, but as an alternative, this object can be computed live during the execution of the app exploiting the <code>goana</code> function, provided by the <code>limma</code> package. Although this likely provides more general (and probably less informative) functions, it is a good compromise for obtaining a further data interpretation.</p></li>
<li><p><code>annotation</code>, a data frame object, with <code>row.names</code> as gene identifiers (e.g. ENSEMBL ids) identical to the row names of the count matrix or <code>dds</code> object, and an extra column <code>gene_name</code>, containing e.g. HGNC-based gene symbols. This can be used for making information extraction easier, as ENSEMBL ids (a usual choice when assigning reads to features) do not provide an immediate readout for which gene they refer to. This can be either passed as a parameter when launching the app, or also uploaded as a tab separated text file.</p></li>
</ul>
</div>
<div id="contact" class="section level2">
<h2 class="hasAnchor">
<a href="#contact" class="anchor"></a>Contact</h2>
<p>For additional details regarding the functions of <strong>pcaExplorer</strong>, please consult the documentation or write an email to <a href="mailto:marinif@uni-mainz.de" class="email">marinif@uni-mainz.de</a>.</p>
<div id="bug-reportsissuesnew-features" class="section level3">
<h3 class="hasAnchor">
<a href="#bug-reportsissuesnew-features" class="anchor"></a>Bug reports/Issues/New features</h3>
<p>Please use <a href="https://github.com/federicomarini/pcaExplorer/issues" class="uri">https://github.com/federicomarini/pcaExplorer/issues</a> for reporting bugs, issues or for suggesting new features to be implemented.</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <div class="links">
<h2>Links</h2>
<ul class="list-unstyled">
<li>Download from BIOC at <br><a href="https://www.bioconductor.org/packages/pcaExplorer">https://​www.bioconductor.org/​packages/​pcaExplorer</a>
</li>
<li>Browse source code at <br><a href="https://github.com/federicomarini/pcaExplorer">https://​github.com/​federicomarini/​pcaExplorer</a>
</li>
<li>Report a bug at <br><a href="https://github.com/federicomarini/pcaExplorer/issues">https://​github.com/​federicomarini/​pcaExplorer/​issues</a>
</li>
</ul>
</div>
<div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li>
<a href="https://opensource.org/licenses/mit-license.php">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a>
</li>
</ul>
</div>
<div class="citation">
<h2>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html">Citing pcaExplorer</a></li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>
<a href="https://federicomarini.github.io">Federico Marini</a> <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0003-3252-7758" target="orcid.widget"><img src="https://members.orcid.org/sites/default/files/vector_iD_icon.svg" class="orcid" alt="ORCID" height="16"></a> </li>
</ul>
</div>

  <div class="dev-status">
<h2>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://codecov.io/github/federicomarini/pcaExplorer?branch=master"><img src="https://codecov.io/github/federicomarini/pcaExplorer/coverage.svg?branch=master" alt="codecov.io"></a></li>
</ul>
</div>
</div>
</div>

      <footer><div class="copyright">
  <p>Developed by <a href="https://federicomarini.github.io">Federico Marini</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
</div>

  

  </body>
</html>
